Autofill credentials in the SAML window from ini file

nitrogl opened the pull request:
This allows one to autofill credentials in the SAML window from ini file.
Default is "~/.gp-saml-gui-credentials", can be specified with "--login" command line argument.

Mode of the credential file should be set the to 600 (similarly to .git-credentials, .ssh/id_rsa, ...).

nitrogl committed:
Read credentials stored in a file

dlenski commented on gp_saml_gui.py:
```diff
@@ -116,6 +131,10 @@ def log_resource_text(self, resource, result, content_type, charset=None, show_h
         if charset or content_type.startswith('text/'):
             print(data.decode(charset or 'utf-8'), file=stderr)
 
+    def setvalue_DOM_element(self, selector, value):
+        if self.wview:
+            self.wview.evaluate_javascript("document.getElementById('" + selector + "').value='" + value + "';", -1, None, None, None, None, None)
```
These need quoting. Using JSON should be sufficient, e.g. `"document.getElementById(%s).value = %s;" % (json.dumps(selector), json.dumps(value));`

nitrogl replied:
I've fixed this one, thank you

dlenski reviewed (commented):
Can the interface here be made compatible with `openconnect --form-entry=NAME:VALUE`? https://www.infradead.org/openconnect/manual.html#opt-form-entry

In particular, form fields should be specified by the _name_ attribute rather than the _id_ attriute.

dlenski commented on gp_saml_gui.py:
```diff
@@ -71,6 +73,19 @@ def __init__(self, uri, html=None, verbose=False, cookies=None, verify=True, use
+                print("Error opening '%s'" % credentials)
+                config = None
+            if config:
+                for x in ['username', 'password']:
+                    self.credentials[x] = config['gp-saml-gui'][x]
```
Why not allow any arbitrarily-named form field to be set? (For example, what if the password field is called `passwd`, what if there's an extra dropdown field called `realm` that needs to be set, etc.)

nitrogl replied:
Not only I've added arbitrarily-named form field to be set, but also domain specific, so in the same file one can store different field-value for different domains: e.g,

```
[domain.one]
user=user
password=password

[domain.two]
email=email
pwd=pwd
```

nitrogl committed:
Merge branch 'master' into master

nitrogl committed:
rename to match argparse Namespace

nitrogl committed:
expand ~

nitrogl committed:
read the ini file content

nitrogl committed:
use domain as section names

nitrogl committed:
fill custom fields

nitrogl committed:
fix issue

dlenski on May 2, 2024:

These need quoting. Using JSON should be sufficient, e.g.
"document.getElementById(%s).value = %s;" % (json.dumps(selector),
json.dumps(value));

nitrogl committed:
add "name" attribute completion

nitrogl commented:
> Can the interface here be made compatible with `openconnect --form-entry=NAME:VALUE`? https://www.infradead.org/openconnect/manual.html#opt-form-entry
> 
> In particular, form fields should be specified by the _name_ attribute rather than the _id_ attriute.

This filling operation happens **before** openconnect is called. Additionally, I'm not sure if we should actually use --form-entry, as the manual suggest "This option should not be used to enter passwords. --passwd-on-stdin should be used for that purpose."

I initially based that on _id_attribute because they're supposed to be unique. To conform to _name_attribute, I've also added code to fill _name_ too! (where _id_ is filled after _name_).
